File: C++/Level2/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

int solution(int k, vector<vector<int>> dungeons) {
    int answer = -1;
    int n = dungeons.size();
    sort(dungeons.begin(), dungeons.end());

    do {
        int count = 0;
        int fatigue = k;

        while (count < n && dungeons[count][0] <= fatigue) 
            fatigue -= dungeons[count++][1];

        answer = max(count, answer);
    } while (next_permutation(dungeons.begin(), dungeons.end()));

    return answer;
}

int main() {
    cout << solution(80, { {80, 20}, {50, 40}, {30, 10} });
    return 0;
}

File: programmers/Java/Level2/피로도.java
class Solution {
    public int solution(int k, int[][] dungeons) {
        return getMaxDungeonCount(k, dungeons, 0, 0);
    }

    public int getMaxDungeonCount(int k, int[][] dungeons, int count, int selected) {
        int maxCount = count;
        for (int i = 0; i < dungeons.length; i++)
            if ((selected & (1 << i)) == 0 && dungeons[i][0] <= k)
                maxCount = Math.max(maxCount, getMaxDungeonCount(k - dungeons[i][1], dungeons, count + 1, selected | (1 << i)));
        return maxCount;
    }

    public static void main(String[] args) {
        System.out.println(1);
    }
}